File: src/lib/trainingService.ts
import { 
  collection, 
  doc, 
  getDoc, 
  getDocs, 
  query, 
  where, 
  addDoc, 
  updateDoc, 
  Timestamp, 
  onSnapshot,
  deleteDoc,
  orderBy
} from 'firebase/firestore';
import { db } from './firebase';
import { TrainingPost, TrainingComment, TrainingApplication } from '../types';

export const trainingService = {
  // Real-time listen of posts
  getPosts(callback: (posts: TrainingPost[]) => void) {
    const q = query(collection(db, 'training_posts'), orderBy('createdAt', 'desc'));
    return onSnapshot(q, (snap) => {
      const posts = snap.docs.map(d => {
        const data = d.data();
        let created;
        if (data.createdAt && typeof data.createdAt.toDate === 'function') {
          created = data.createdAt.toDate().toISOString();
        } else if (data.createdAt && data.createdAt.seconds) {
          created = new Date(data.createdAt.seconds * 1000).toISOString();
        } else {
          created = new Date().toISOString();
        }
        return {
          id: d.id,
          ...data,
          createdAt: created,
        } as TrainingPost;
      });
      callback(posts);
    }, (error) => {
      console.error("Error getting posts: ", error);
    });
  },

  // Create post (normal or training)
  async createPost(post: Omit<TrainingPost, 'id' | 'createdAt' | 'likes' | 'commentsCount'>) {
    const docRef = await addDoc(collection(db, 'training_posts'), {
      ...post,
      likes: [],
      commentsCount: 0,
      createdAt: Timestamp.now()
    });
    
    // If it's a training post, notify staff
    if (post.isTraining) {
      const teachersSnap = await getDocs(query(collection(db, 'teachers'), where('status', '==', 'active')));
      for (const tDoc of teachersSnap.docs) {
        await addDoc(collection(db, 'notifications'), {
          userId: tDoc.id,
          title: `📖 New Training Available`,
          message: `The Principal posted a new training: "${post.trainingTitle}". Apply now!`,
          read: false,
          timestamp: Timestamp.now(),
          type: 'admin'
        });
      }
    }
    return docRef.id;
  },

  // Highlight font styles
  fontFamilies: {
    sans: 'font-sans',
    serif: 'font-serif',
    mono: 'font-mono',
    playful: 'font-sans font-medium tracking-normal italic text-indigo-700',
    elegant: 'font-serif italic font-semibold text-[#8B7E6D]'
  },

  // Like / Unlike post
  async toggleLike(postId: string, userId: string) {
    const postRef = doc(db, 'training_posts', postId);
    const snap = await getDoc(postRef);
    if (!snap.exists()) return;
    const data = snap.data();
    const likes = data.likes || [];
    const newLikes = likes.includes(userId)
      ? likes.filter((id: string) => id !== userId)
      : [...likes, userId];
    
    await updateDoc(postRef, { likes: newLikes });
  },

  // Comments
  getComments(postId: string, callback: (comments: TrainingComment[]) => void) {
    const q = query(
      collection(db, 'training_comments'), 
      where('postId', '==', postId),
      orderBy('createdAt', 'asc')
    );
    return onSnapshot(q, (snap) => {
      const comments = snap.docs.map(d => {
        const data = d.data();
        let created;
        if (data.createdAt && typeof data.createdAt.toDate === 'function') {
          created = data.createdAt.toDate().toISOString();
        } else if (data.createdAt && data.createdAt.seconds) {
          created = new Date(data.createdAt.seconds * 1000).toISOString();
        } else {
          created = new Date().toISOString();
        }
        return {
          id: d.id,
          ...data,
          createdAt: created,
        } as TrainingComment;
      });
      callback(comments);
    });
  },

  async addComment(postId: string, comment: Omit<TrainingComment, 'id' | 'postId' | 'createdAt'>) {
    await addDoc(collection(db, 'training_comments'), {
      ...comment,
      postId,
      createdAt: Timestamp.now()
    });

    // Update comments count on post
    const postRef = doc(db, 'training_posts', postId);
    const postSnap = await getDoc(postRef);
    if (postSnap.exists()) {
      const currentCount = postSnap.data().commentsCount || 0;
      await updateDoc(postRef, { commentsCount: currentCount + 1 });
    }
  },

  // Applications
  async applyForTraining(postId: string, teacherId: string, teacherName: string, trainingTitle: string) {
    // Check if employee already applied
    const q = query(
      collection(db, 'training_applications'),
      where('postId', '==', postId),
      where('teacherId', '==', teacherId)
    );
    const snap = await getDocs(q);
    if (!snap.empty) {
      throw new Error(`You have already applied for this training.`);
    }

    // Submit training application
    const docRef = await addDoc(collection(db, 'training_applications'), {
      postId,
      trainingTitle,
      teacherId,
      teacherName,
      status: 'pending',
      createdAt: Timestamp.now()
    });

    // Add alert notification for Principal / admin
    await addDoc(collection(db, 'notifications'), {
      userId: 'admin-id',
      title: `📝 Training Application: ${teacherName}`,
      message: `${teacherName} applied for "${trainingTitle}" training session.`,
      read: false,
      timestamp: Timestamp.now(),
      type: 'leave' // Using leave/admin context for approval monitoring
    });

    return docRef.id;
  },

  // Real-time sync of applications
  getApplications(callback: (applications: TrainingApplication[]) => void) {
    const q = query(collection(db, 'training_applications'), orderBy('createdAt', 'desc'));
    return onSnapshot(q, (snap) => {
      const applications = snap.docs.map(d => {
        const data = d.data();
        let created;
        if (data.createdAt && typeof data.createdAt.toDate === 'function') {
          created = data.createdAt.toDate().toISOString();
        } else if (data.createdAt && data.createdAt.seconds) {
          created = new Date(data.createdAt.seconds * 1000).toISOString();
        } else {
          created = new Date().toISOString();
        }
        return {
          id: d.id,
          ...data,
          createdAt: created,
        } as TrainingApplication;
      });
      callback(applications);
    }, (error) => {
      console.error("Error reading applications: ", error);
    });
  },

  // Update application status (Approve or Disapprove)
  async updateApplicationStatus(applicationId: string, postId: string, status: 'approved' | 'rejected', teacherId: string, trainingTitle: string) {
    await updateDoc(doc(db, 'training_applications', applicationId), { status });

    // Update authorized trainees list on post if approved
    if (status === 'approved') {
      const postRef = doc(db, 'training_posts', postId);
      const postSnap = await getDoc(postRef);
      if (postSnap.exists()) {
        const data = postSnap.data();
        const traineeIds = data.traineeIds || [];
        if (!traineeIds.includes(teacherId)) {
          await updateDoc(postRef, {
            traineeIds: [...traineeIds, teacherId]
          });
        }
      }
    }

    // Send visual notification update directly to the teacher
    await addDoc(collection(db, 'notifications'), {
      userId: teacherId,
      title: status === 'approved' ? `✅ Training Approved!` : `❌ Training Application Rejected`,
      message: `Your training application for "${trainingTitle}" has been ${status === 'approved' ? 'approved' : 'rejected'} by the administration.`,
      read: false,
      timestamp: Timestamp.now(),
      type: 'admin'
    });
  },

  // Check if a teacher has already applied
  async getTeacherApplicationStatus(postId: string, teacherId: string): Promise<'pending' | 'approved' | 'rejected' | null> {
    const q = query(
      collection(db, 'training_applications'),
      where('postId', '==', postId),
      where('teacherId', '==', teacherId)
    );
    const snap = await getDocs(q);
    if (snap.empty) return null;
    return snap.docs[0].data().status as any;
  }
};

File: lib/modules/training/services/training_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../teachers/models/teacher.dart';
import '../models/training.dart';

class TrainingService {
  TrainingService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _db;

  CollectionReference<Map<String, dynamic>> get _posts =>
      _db.collection('training_posts');
  CollectionReference<Map<String, dynamic>> get _applications =>
      _db.collection('trainingApplications');
  CollectionReference<Map<String, dynamic>> get _teachers =>
      _db.collection('teachers');

  Future<String> createPost(TrainingPost post) async {
    final doc = post.id.isEmpty ? _posts.doc() : _posts.doc(post.id);
    await doc.set(post.toMap());
    return doc.id;
  }

  Stream<List<TrainingPost>> streamPosts({bool trainingOnly = false}) {
    Query<Map<String, dynamic>> query =
        _posts.orderBy('createdAt', descending: true);
    if (trainingOnly) {
      query = query.where('isTraining', isEqualTo: true);
    }

    return query.snapshots().map(
          (snapshot) => snapshot.docs
              .map((doc) => TrainingPost.fromMap(doc.id, doc.data()))
              .toList(),
        );
  }

  Stream<List<TrainingPost>> getTrainingPosts() => streamPosts();

  Future<void> toggleLike(String postId, String userId) async {
    final ref = _posts.doc(postId);
    await _db.runTransaction((transaction) async {
      final snapshot = await transaction.get(ref);
      if (!snapshot.exists) return;

      final likes = List<String>.from(snapshot.data()?['likes'] ?? []);
      transaction.update(ref, {
        'likes': likes.contains(userId)
            ? FieldValue.arrayRemove([userId])
            : FieldValue.arrayUnion([userId]),
      });
    });
  }

  Future<void> addComment({
    required String postId,
    required String authorId,
    required String authorName,
    required String authorRole,
    required String text,
  }) async {
    final postRef = _posts.doc(postId);
    final commentRef = postRef.collection('comments').doc();
    final comment = TrainingComment(
      id: commentRef.id,
      postId: postId,
      authorId: authorId,
      authorName: authorName,
      authorRole: authorRole,
      text: text.trim(),
      createdAt: DateTime.now(),
    );

    await _db.runTransaction((transaction) async {
      transaction.set(commentRef, comment.toMap());
      transaction.update(postRef, {'commentsCount': FieldValue.increment(1)});
    });
  }

  Stream<List<TrainingComment>> streamComments(String postId) {
    return _posts
        .doc(postId)
        .collection('comments')
        .orderBy('createdAt')
        .snapshots()
        .map((snapshot) => snapshot.docs
            .map((doc) => TrainingComment.fromMap(doc.id, doc.data()))
            .toList());
  }

  Future<void> applyTraining(TrainingApplication application) async {
    final appRef =
        _applications.doc('${application.postId}_${application.teacherId}');
    final postRef = _posts.doc(application.postId);

    await _db.runTransaction((transaction) async {
      final postSnapshot = await transaction.get(postRef);
      if (!postSnapshot.exists) {
        throw StateError('Training post no longer exists.');
      }

      final post = TrainingPost.fromMap(postSnapshot.id, postSnapshot.data()!);
      if (!post.isTraining || !post.isOpenVolunteer) {
        throw StateError(
            'This training is not open for volunteer applications.');
      }
      if (post.isFull) {
        throw StateError('This training is already full.');
      }
      if (post.traineeIds.contains(application.teacherId)) {
        throw StateError('You are already enrolled in this training.');
      }

      final appSnapshot = await transaction.get(appRef);
      if (appSnapshot.exists) {
        final status = appSnapshot.data()?['status'] ?? 'pending';
        if (status == 'pending' || status == 'approved') {
          throw StateError(
              'You already have an application for this training.');
        }
      }

      transaction.set(appRef, application.toMap());
    });
  }

  Future<void> approveApplication(TrainingApplication application) async {
    final appRef = _applications.doc(application.id);
    final postRef = _posts.doc(application.postId);

    await _db.runTransaction((transaction) async {
      final postSnapshot = await transaction.get(postRef);
      if (!postSnapshot.exists) {
        throw StateError('Training post no longer exists.');
      }

      final post = TrainingPost.fromMap(postSnapshot.id, postSnapshot.data()!);
      if (post.isFull && !post.traineeIds.contains(application.teacherId)) {
        throw StateError('No seats remain for this training.');
      }

      transaction.update(appRef, {'status': 'approved'});
      transaction.update(postRef, {
        'traineeIds': FieldValue.arrayUnion([application.teacherId]),
      });
    });
  }

  Future<void> rejectApplication(String applicationId) async {
    await _applications.doc(applicationId).update({'status': 'rejected'});
  }

  Future<void> assignTraineeToTraining({
    required String postId,
    required String teacherId,
  }) async {
    final postRef = _posts.doc(postId);
    await _db.runTransaction((transaction) async {
      final postSnapshot = await transaction.get(postRef);
      if (!postSnapshot.exists) {
        throw StateError('Training post no longer exists.');
      }

      final post = TrainingPost.fromMap(postSnapshot.id, postSnapshot.data()!);
      if (post.isFull && !post.traineeIds.contains(teacherId)) {
        throw StateError('No seats remain for this training.');
      }

      transaction.update(postRef, {
        'traineeIds': FieldValue.arrayUnion([teacherId]),
      });
    });
  }

  Stream<List<TrainingApplication>> streamApplications({
    String? status,
    String? teacherId,
    String? postId,
  }) {
    Query<Map<String, dynamic>> query =
        _applications.orderBy('createdAt', descending: true);
    if (status != null) {
      query = query.where('status', isEqualTo: status);
    }
    if (teacherId != null) {
      query = query.where('teacherId', isEqualTo: teacherId);
    }
    if (postId != null) {
      query = query.where('postId', isEqualTo: postId);
    }

    return query.snapshots().map(
          (snapshot) => snapshot.docs
              .map((doc) => TrainingApplication.fromMap(doc.id, doc.data()))
              .toList(),
        );
  }

  Future<TeacherRecord?> getFacultyProfile(String authorId) async {
    final doc = await _teachers.doc(authorId).get();
    if (!doc.exists || doc.data() == null) return null;
    return TeacherRecord.fromMap(doc.id, doc.data()!);
  }
}
